## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [492]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [493]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [494]:
# Use the describe function to gather some basic statistics
df.shape

(54639, 18)

In [495]:
# Use this space to do any additional research
# and familiarize yourself with the data.
renamed_df = df.rename(columns={'first': 'First Name', 
                                'last': 'Last Name', 
                                'job': 'Job', 
                                'phone': 'Phone', 
                                'email': 'Email', 
                                'client_id' : 'Client ID', 
                                'order_id' : 'Order ID', 
                                'order_date' : 'Order Date', 
                                'order_week' : 'Order Week',
                                'order_year' : 'Order Year',
                                'item_id' : 'Item ID',
                                'category' : 'Category',
                                'subcategory' : 'Subcategory',
                                'unit_price' : 'Unit Price',
                                'unit_cost' : 'Unit Cost',
                                'unit_weight' : 'Unit Weight',
                                'qty' : 'Quantity',
                                'line_number' : 'Line Number'})
renamed_df.count()

First Name     54639
Last Name      54639
Job            54639
Phone          54639
Email          54639
Client ID      54639
Order ID       54639
Order Date     54639
Order Week     54639
Order Year     54639
Item ID        54639
Category       54639
Subcategory    54639
Unit Price     54639
Unit Cost      54639
Unit Weight    54639
Quantity       54639
Line Number    54639
dtype: int64

In [496]:
# What three item categories had the most entries?

categories_df = renamed_df["Category"].value_counts()
categories_df.head(3)
#top_3_category_names = top_3_categories.index
#print(f'The top three categories with the most entries are {top_3_category_names.tolist()}.')


Category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [497]:
# For the category with the most entries, which subcategory had the most entries?
#consumable_subcategory_df = renamed_df.set_index(renamed_df['Category'], inplace= True)
#categories_df = renamed_df.sort_index()

#top_category_subcategory_df = categories_df.set_index(categories_df['Category'], inplace= True)
#top_category_subcategory_df = categories_df.sort_index()
top_category = renamed_df["Category"].value_counts().index[0]
top_category_df = renamed_df[renamed_df['Category'] == top_category]
top_category_df["Subcategory"].value_counts().head(1)
#top_category.head(1)


#consumable_entries = consumables_df["Subcategory"].value_counts()
#top_consumable_subcategory = consumable_entries.head(1)
#top_consumable_subcategory_name = top_consumable_subcategory.index
#print(f'The subcategory with the most entries, in the category with the most entries (consumables), is {top_consumable_subcategory_name.to_list()}.')

Subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [498]:
# Which five clients had the most entries in the data?

renamed_df['Client ID'].value_counts().head()

#client_ranked_df = renamed_df.sort_values(by= 'Quantity', ascending= False)
#client_ranked_df = client_ranked_df.reset_index(drop=True)

#client_ranked_df.head()

#print(f'The top five clients with most entries are {top_5_client_names.to_list()}.')

Client ID
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [499]:
# Store the client ids of those top 5 clients in a list.
top_client_ids = renamed_df['Client ID'].value_counts().head(5).index.to_list()
print(top_client_ids)
#top_client_ids


[33615, 66037, 46820, 38378, 24741]


In [512]:
# How many total units (the qty column) did the client with the most entries order order?
top_client_total_units = client_ranked_df.iloc[0, 16]

print(f'The top client ordered {top_client_total_units} units.')

The top client ordered 3958244 units.


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [501]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty

subtotal = renamed_df['Unit Price'] * renamed_df['Quantity']
renamed_df['Subtotal'] = subtotal

renamed_df.head()

First Name Last Name                    Job               Phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 Email  Client ID  Order ID  Order Date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   Order Week  Order Year         Item ID     Category         Subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   Unit Price  Unit Cost  Unit Weight  Quantity  Line Number   Subtotal  
0     1096.80     762.71         7.50       105            1  115164.00  
1       24.95      15.09         1.49        21            0     523.95  
2       13.52       7.86         1.68        39            6     527.28  
3       36.42      24.85         1.23        29            3    1056.18  
4      195.10     108.17        46.43        20            1    3902.00

In [502]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
def shipping_price (row):
    if (row['Unit Weight'] < 50):
        return row['Unit Weight'] * 10
    else:
        return row['Unit Weight'] * 7
    
renamed_df['Shipping Price'] = renamed_df.apply(shipping_price, axis = 1)
renamed_df.head()
        



First Name Last Name                    Job               Phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 Email  Client ID  Order ID  Order Date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   Order Week  Order Year         Item ID     Category         Subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   Unit Price  Unit Cost  Unit Weight  Quantity  Line Number   Subtotal  \
0     1096.80     762.71         7.50       105            1  115164.00   
1       24.95      15.09         1.49        21            0     523.95   
2       13.52       7.86         1.68        39            6     527.28   
3       36.42      24.85         1.23        29            3    1056.18   
4      195.10     108.17        46.43        20            1    3902.00   

   Shipping Price  
0            75.0  
1            14.9  
2            16.8  
3            12.3  
4           464.3

In [503]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
total_price = (renamed_df['Shipping Price'] + renamed_df['Subtotal']) * 0.0925
renamed_df['Line Price'] = total_price

renamed_df.tail()

First Name  Last Name                      Job               Phone  \
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                    Email  Client ID  Order ID  Order Date  \
54634  late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635        derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636           gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637           solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                    belld@galvan.com      24007   7489403  2023-01-04   

       Order Week  Order Year  ...     Category         Subcategory  \
54634          17        2023  ...     software           logistics   
54635           6        2023  ...  consumables    kitchen supplies   
54636           9        2023  ...  consumables   bathroom supplies   
54637           3        2023  ...    furniture              stands   
54638           1        2023  ...     software  project management   

      Unit Price  Unit Cost  Unit Weight  Quantity  Line Number  Subtotal  \
54634      83.13      51.60         2.25        33            8   2743.29   
54635     206.59     175.46        11.70        47            0   9709.73   
54636      65.66      57.31         4.16       475            6  31188.50   
54637       1.48       1.22        18.04       112            0    165.76   
54638       3.01       1.61         2.07      1031            5   3103.31   

       Shipping Price   Line Price  
54634            22.5   255.835575  
54635           117.0   908.972525  
54636            41.6  2888.784250  
54637           180.4    32.019800  
54638            20.7   288.970925  

[5 rows x 21 columns]

In [504]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
line_cost = (renamed_df['Unit Cost'] * renamed_df['Quantity']) + renamed_df['Shipping Price']
renamed_df['Line Cost'] = line_cost

renamed_df.head()


First Name Last Name                    Job               Phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 Email  Client ID  Order ID  Order Date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   Order Week  Order Year  ...         Subcategory Unit Price Unit Cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   Unit Weight  Quantity  Line Number   Subtotal  Shipping Price  \
0         7.50       105            1  115164.00            75.0   
1         1.49        21            0     523.95            14.9   
2         1.68        39            6     527.28            16.8   
3         1.23        29            3    1056.18            12.3   
4        46.43        20            1    3902.00           464.3   

     Line Price  Line Cost  
0  10659.607500   80159.55  
1     49.843625     331.79  
2     50.327400     323.34  
3     98.834400     732.95  
4    403.882750    2627.70  

[5 rows x 22 columns]

In [505]:
# Create a column for the profit of each line using line cost and line price
line_profit = renamed_df['Line Price'] - renamed_df['Line Cost']
renamed_df['Profit'] = line_profit

renamed_df.head()


First Name Last Name                    Job               Phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 Email  Client ID  Order ID  Order Date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   Order Week  Order Year  ... Unit Price Unit Cost Unit Weight  Quantity  \
0          17        2023  ...    1096.80    762.71        7.50       105   
1          20        2023  ...      24.95     15.09        1.49        21   
2           4        2023  ...      13.52      7.86        1.68        39   
3          17        2023  ...      36.42     24.85        1.23        29   
4          21        2023  ...     195.10    108.17       46.43        20   

   Line Number   Subtotal  Shipping Price    Line Price  Line Cost  \
0            1  115164.00            75.0  10659.607500   80159.55   
1            0     523.95            14.9     49.843625     331.79   
2            6     527.28            16.8     50.327400     323.34   
3            3    1056.18            12.3     98.834400     732.95   
4            1    3902.00           464.3    403.882750    2627.70   

         Profit  
0 -69499.942500  
1   -281.946375  
2   -273.012600  
3   -634.115600  
4  -2223.817250  

[5 rows x 23 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [506]:
# Check your work using the totals above
#order_id_df = client_ranked_df.sort_values(by= 'Order ID', ascending= True)
#order_id_df = client_ranked_df.reset_index(drop=True)
#order_2742071 = renamed_df[renamed_df['Order ID'] == '2742071']
#print(order_2742071)
#fm16_df = renamed_df.loc[renamed_df['report_type'] == 'FM-16']
#len(fm16_df)

def order_check (order_id):
    order_total = renamed_df[renamed_df['Order ID'] == order_id]['Line Price'].sum()
    print(f'The total price for the order is ${order_total}.')
    return

renamed_df

First Name  Last Name                      Job               Phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     Email  Client ID  Order ID  Order Date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       Order Week  Order Year  ... Unit Price Unit Cost Unit Weight  Quantity  \
0              17        2023  ...    1096.80    762.71        7.50       105   
1              20        2023  ...      24.95     15.09        1.49        21   
2               4        2023  ...      13.52      7.86        1.68        39   
3              17        2023  ...      36.42     24.85        1.23        29   
4              21        2023  ...     195.10    108.17       46.43        20   
...           ...         ...  ...        ...       ...         ...       ...   
54634          17        2023  ...      83.13     51.60        2.25        33   
54635           6        2023  ...     206.59    175.46       11.70        47   
54636           9        2023  ...      65.66     57.31        4.16       475   
54637           3        2023  ...       1.48      1.22       18.04       112   
54638           1        2023  ...       3.01      1.61        2.07      1031   

       Line Number   Subtotal  Shipping Price    Line Price  Line Cost  \
0                1  115164.00            75.0  10659.607500   80159.55   
1                0     523.95            14.9     49.843625     331.79   
2                6     527.28            16.8     50.327400     323.34   
3                3    1056.18            12.3     98.834400     732.95   
4                1    3902.00           464.3    403.882750    2627.70   
...            ...        ...             ...           ...        ...   
54634            8    2743.29            22.5    255.835575    1725.30   
54635            0    9709.73           117.0    908.972525    8363.62   
54636            6   31188.50            41.6   2888.784250   27263.85   
54637            0     165.76           180.4     32.019800     317.04   
54638            5    3103.31            20.7    288.970925    1680.61   

             Profit  
0     -69499.942500  
1       -281.946375  
2       -273.012600  
3       -634.115600  
4      -2223.817250  
...             ...  
54634  -1469.464425  
54635  -7454.647475  
54636 -24375.065750  
54637   -285.020200  
54638  -1391.639075  

[54639 rows x 23 

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [507]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.



In [508]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 


In [509]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 


# Define a function that converts a dollar amount to millions.


# Apply the currency_format_millions function to only the money columns. 


# Rename the columns to reflect the change in the money format. 




In [510]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
